--[[
SPDX-License-Identifier: MPL-2.0
]]

local profiles = {}

profiles.core = {
	"data-and-control-flow/nil.lsp",
	"data-and-control-flow/t.lsp",
	"data-and-control-flow/if.lsp",
	"data-and-control-flow/progn.lsp",
	"data-and-control-flow/and.lsp",
	"data-and-control-flow/or.lsp",
	"data-and-control-flow/when.lsp",
	"data-and-control-flow/unless.lsp",
	"data-and-control-flow/let.lsp",
	"data-and-control-flow/letstar.lsp",
	"data-and-control-flow/defparameter.lsp",
	"data-and-control-flow/defvar.lsp",
	"data-and-control-flow/defun.lsp",
}

profiles.clos = {
	"objects/add-method.lsp",
	"objects/allocate-instance.lsp",
	"objects/call-next-method.lsp",
	"objects/change-class.lsp",
	"objects/class-name.lsp",
	"objects/class-of.lsp",
	"objects/compute-applicable-methods.lsp",
	"objects/defclass-01.lsp",
	"objects/defclass-02.lsp",
	"objects/defclass-03.lsp",
	"objects/defclass-errors.lsp",
	"objects/defclass-forward-reference.lsp",
	"objects/defclass.lsp",
	"objects/defgeneric-method-combination-and.lsp",
	"objects/defgeneric-method-combination-append.lsp",
	"objects/defgeneric-method-combination-aux.lsp",
	"objects/defgeneric-method-combination-list.lsp",
	"objects/defgeneric-method-combination-max.lsp",
	"objects/defgeneric-method-combination-min.lsp",
	"objects/defgeneric-method-combination-nconc.lsp",
	"objects/defgeneric-method-combination-or.lsp",
	"objects/defgeneric-method-combination-plus.lsp",
	"objects/defgeneric-method-combination-progn.lsp",
	"objects/defgeneric.lsp",
	"objects/define-method-combination-long-form.lsp",
	"objects/define-method-combination.lsp",
	"objects/defmethod.lsp",
	"objects/ensure-generic-function.lsp",
	"objects/find-class.lsp",
	"objects/find-method.lsp",
	"objects/load.lsp",
	"objects/make-instance.lsp",
	"objects/make-instances-obsolete.lsp",
	"objects/make-load-form-order.lsp",
	"objects/make-load-form-saving-slots.lsp",
	"objects/make-load-form.lsp",
	"objects/method-qualifiers.lsp",
	"objects/next-method-p.lsp",
	"objects/no-applicable-method.lsp",
	"objects/no-next-method.lsp",
	"objects/reinitialize-instance.lsp",
	"objects/remove-method.lsp",
	"objects/shared-initialize.lsp",
	"objects/slot-boundp.lsp",
	"objects/slot-exists-p.lsp",
	"objects/slot-makunbound.lsp",
	"objects/slot-missing.lsp",
	"objects/slot-unbound.lsp",
	"objects/slot-value.lsp",
	"objects/unbound-slot.lsp",
	"objects/update-instance-for-different-class.lsp",
	"objects/with-accessors.lsp",
	"objects/with-slots.lsp",
	"types-and-classes/class-precedence-lists.lsp",
	"types-and-classes/standard-generic-function.lsp",
	"types-and-classes/type-of.lsp",
	"types-and-classes/types-and-class.lsp",
	"types-and-classes/types-and-class-2.lsp",
}

return profiles
